rfcs/react-components/convergence/portal-compat.md

```diff
@@ -0,0 +1,166 @@
+
+In the prototype the publisher component is called `CompatHostV9`:
+
+- it reads current classes from `useThemeClassName()`
+- adds `fui-FluentProvider*` class to any portal element that was registered by subscribers
+
+Functional prototype: https://codesandbox.io/s/portal-compat-prototype-f59rn
+
+#### Changes to v0 & v8 code
+
+`PortalInner` in `@fluentui/react-northstar` & `Layer` in `@fluentui/react` should:
+
+- subscribe for new context
+- register elements that are passed to `React.createPortal()` as `container`
+
+Sample code is below:
+
+```js
+import React from 'react';
+
+// "PortalCompatContext" & "usePortalCompat()" hook is everything that will be included to v8 & v0
+
+export const PortalCompatContext = React.createContext(
+  () => {},
+  () => {},
+);
+
+export function usePortalCompat() {
+  return React.useContext(PortalCompatContext);
+}
+
+function Portal() {
+  // ⚠️ "document.createElement" is used for demo purposes
+  //   This should be a real element where a portal will be mounted
+  const portalElement = document.createElement('div');
+  const [registerPortalEl, unregisterPortalEl] = usePortalCompat();
+
+  React.useEffect(() => {
+    registerPortalEl(portalElement);
+
+    return () => {
+      unregisterPortalEl(portalElement);
+    };
+  }, [portalElement, registerPortalEl, unregisterPortalEl]);
+}
+```
+
+With these changes v0 & v8 portals will get proper classes with variables ➡️ styles for v9 components will look properly.
+
+#### How this will be used?
+
+Once changes to v0 & v8 will be implemented customers would be able to inject `CompatHostV9` (should be a child of `FluentProvider` from v9):
+
+```tsx
+<V9Provider>
+  {/* 👇 Needs a better name */}
+  <CompatHostV9>
+    <V0Provider>{/* Components from v9 or v0... */}</V0Provider>
+  </CompatHostV9>
+</V9Provider>
+```
+
+As there might be customers that use only v9 it's reasonable to keep `CompatHostV9` separated from `V9Provider`.
+
+#### Where "compat" package lives?
+
+As this package will be a dependency of both v0 & v8 we will get it into the dependency graph:
+
+```
+                 -----------------------------------
+                 | @fluentui/react-portal-compat@9 |
+                 -----------------------------------
+                      ▲                       ▲
+                     /                         \
+                    /                           \
+                   /                             \
+                  /                               \
+                 /                                 \
+                /                                   \
+  ---------------------                        -------------------------------
+  | @fluentui/react@8 |                        | @fluentui/react-northstar@0 |
+  ---------------------                        -------------------------------
+
+```
+
+The suggestion is to keep it inside monorepo and put it into a new subdirectory `/packages/shared/` or `/packages/compat/`.
```

Comment:
@GeoffCox  is working on migration and compat, and has a separate repo for tools like this, should it live there instead of the main Fluent repo ?

Reply:
My experience with OSS is that it is not common practice to make compatware like this. We would be doing this explicitly for our enterprise partners to avoid upgrade from v8->v9

Reply:
rxjs uses compat layering for example. 

From domain boundary it makes perfect sense to have this in our repo as it's "strictly coupled" (unlike griffel for instance)

> avoid upgrade ? 

I don't follow.

Reply:
resolving this, it's not a blocker in any case, and there are more important things to consider for eventual release :)